subsys/mpsl/fem/nrf2220/mpsl_fem_nrf2220.c

```diff
@@ -193,4 +206,139 @@ BUILD_ASSERT(CONFIG_MPSL_FEM_INIT_PRIORITY > CONFIG_I2C_INIT_PRIORITY,
 
 SYS_INIT(mpsl_fem_init, POST_KERNEL, CONFIG_MPSL_FEM_INIT_PRIORITY);
 
+#if defined(CONFIG_MPSL_FEM_NRF2220_TEMPERATURE_COMPENSATION)
+
+static K_SEM_DEFINE(fem_temperature_new_value_sem, 0, 1);
+
+#if defined(CONFIG_MPSL_FEM_NRF2220_TEMPERATURE_COMPENSATION_WITH_MPSL_SCHEDULER)
+static int32_t fem_temperature_update_result;
+static K_SEM_DEFINE(fem_temperature_updated_cb_sem, 0, 1);
+
+static void fem_temperature_update_cb(int32_t res)
+{
+	fem_temperature_update_result = res;
+	k_sem_give(&fem_temperature_updated_cb_sem);
+}
+
+static int32_t fem_temperature_changed_update_now(void)
+{
+	const struct device * i2c_bus_dev = DEVICE_DT_GET(DT_BUS(MPSL_FEM_TWI_IF));
+
+	/* Let's have initially "taken" semaphore that will inform the operation on the nrf2220 is finished. */
+	k_sem_take(&fem_temperature_updated_cb_sem, K_NO_WAIT);
+
+	(void)i2c_nrfx_twim_exclusive_access_acquire(i2c_bus_dev, K_FOREVER);
+
+	/* Temporary raise i2c_bus_dev IRQ priority, as required by the
+	 * mpsl_fem_nrf2220_temperature_changed_update_request function.
+	 * The IRQs from i2c may not be delayed while performing operations
+	 * in a timeslot granted by the MPSL scheduler.
+	 */
+	z_arm_irq_priority_set(DT_IRQN(DT_BUS(MPSL_FEM_TWI_IF)), 0, IRQ_ZERO_LATENCY);
```

Comment:
My only worry with boosting the priority to ZLI is that the I2C driver still uses the Zephyr's `isr_wrapper`, which expects to run on normal interrupt priorities. Changing this might break the wrapper's assumptions and lead to faults. Could you ask someone who knows the wrapper implementation whether this is okay to do?

Reply:
Please have a look on [the Zephyr doc](https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#:~:text=In%20addition%20to%20that%2C%20the%20flag%20IRQ_ZERO_LATENCY%20must%20be%20passed%20to%20IRQ_CONNECT%20or%20IRQ_DIRECT_CONNECT%20macros%20to%20configure%20the%20particular%20interrupt%20with%20zero%20latency.)
Excerpt from it

_In addition to that, the flag IRQ_ZERO_LATENCY must be passed to IRQ_CONNECT or IRQ_DIRECT_CONNECT macros to configure the particular interrupt with zero latency._

Reply:
> So, based on the doc above, if IRQ_ZERO_LATENCY can be used with IRQ_CONNECT, which introduces the _isr_wrapper then _isr_wrapper is safe to be used in ZLIs.

Sounds like that might be indeed the case. Interrupts created with `IRQ_CONNECT` can also be zero latency IRQs. I have looked into the code and it doesn't look like the `flags_p` parameter affects the  `isr_wrapper` in some way.